src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,375 @@
+        , touchedVersion INT ** maybe
+        )
+
+  tableid (primary key)
+
+  touched table
+
+
+
+-}
+
+detectVersionChange ::
+     Database
+  -> BlockHeight
+  -> BlockHash
+  -> IO (Either String OnVersionChange)
+detectVersionChange c bRestore hsh  = do
+  let sqlBtoBH [[SInt bh]] = BlockHeight (fromIntegral bh)
+      sqlBtoBH _ = error "sqlBtoBH: wat"
+      sqlVtoV [[SInt version]] = ReorgVersion version
+      sqlVtoV _ = error "sqlVtoV: wat"
+  bCurrent <- sqlBtoBH <$>
+    qry_ c "SELECT max(blockheight) AS current_block_height FROM BlockHistory;" [RInt]
+  vCurrent <- sqlVtoV <$>
+    qry_ c "SELECT max(version) AS current_version FROM VersionHistory;" [RInt]
+
+  -- enforce invariant that the history has (B_restore-1,H_parent).
+  historyInvariant <-
+    qry c "SELECT COUNT(*)\
+           \ FROM BlockHistory\
+           \ WHERE blockheight = (?)\
+           \ AND hash = (?);" [SInt $ fromIntegral $ pred bRestore, SBlob (encode hsh)] [RInt]
+
+  if historyInvariant /= [[SInt 1]]
+    then return $ Left "HistoryInvariantViolation"
+  -- enforce invariant that B_restore is not greater than B_current + 1
+    else return $ case compare bRestore (bCurrent + 1) of
+        GT -> Left "RestoreInvariantViolation"
+        EQ -> Right $ NormalOperation (BlockVersion bRestore vCurrent)
+        LT -> Right $ Forking (BlockVersion bRestore (vCurrent + 1))
+
+data OnVersionChange
+  = Forking !BlockVersion
+  | NormalOperation !BlockVersion
+
+onVersionChange ::
+     Database
+  -> BlockHeight
+  -> BlockVersion
+  -> IO (Either String (BlockVersion, SnapshotName))
+onVersionChange c bRestore (BlockVersion bCurrent vCurrent) =
+  case compare bRestore (bCurrent + 1) of
+
+    -- Probably better to throw an exception here.
+    GT -> return $ Left "bRestore should not be greater than bCurrent + 1"
+
+    -- normal operation
+    EQ -> do
+      name <- mkSnapshot c
+      -- set environment blockVersion
+      let bv_env = BlockVersion bRestore vCurrent
+      return $ Right (bv_env, name)
+
+    LT -> do
+      let bv_env = BlockVersion bRestore (vCurrent + 1)
+      _tableMaintenanceRowsVersionedTables c bv_env
+      _tableMaintenanceVersionedTables c bv_env
+      _deleteHistory c bv_env
+      name <- mkSnapshot c
+      return $ Right (bv_env, name)
+
+_tableMaintenanceRowsVersionedTables :: Database -> BlockVersion -> IO ()
+_tableMaintenanceRowsVersionedTables _c _bv = undefined
+
+_tableMaintenanceVersionedTables :: Database -> BlockVersion -> IO ()
+_tableMaintenanceVersionedTables _c _bv = undefined
+
+_deleteHistory :: Database -> BlockVersion -> IO ()
+_deleteHistory _c _bv = do
+  undefined
+
+mkSnapshot :: Database -> IO SnapshotName
+mkSnapshot c = do
+  s@(SnapshotName name) <- mkSnapshotName
+  exec' c "SAVEPOINT (?)" [SBlob name]
+  return s
+
+newtype SnapshotName = SnapshotName BS.ByteString
+  deriving (Eq, Ord, IsString)
+  deriving newtype Show
+
+mkSnapshotName :: IO SnapshotName
```

Comment:
snapshots don't actually need unique names, and re-using one can help enforce that you're not overlapping them. Also they're not snapshots, they're savepoints, so name accordingly:
- for the block savepoint, just call it "BLOCK"
- for each transaction savepoint, call it "TX"
Again, neither of these can overlap at their respective levels, so unique names are unnecessary.

Reply:
I'm not sure how snapshot slipped in there.